File: ClassesDeUmJogo/Csharp/Entidades/Heroi.cs
namespace Csharp.Entidades
{
    public abstract class Heroi
    {
        public string Nome { get; } = default!;
        public int Idade { get; } = default!;
        public string Tipo { get; } = default!;
        public string Ataque { get; } = default!;

        public Heroi(string nome, int idade, string tipo, string ataque)
        {
            Nome = nome;
            Idade = idade;
            Tipo = tipo;
            Ataque = ataque;
        }

        public void Atacar() {
            Console.WriteLine($"O {Tipo} atacou usando {Ataque}.");
        }
    }
}

File: ClassesDeUmJogo/Javascript/Entidades/Heroi.js
class Heroi {
    constructor(nome, idade, tipo, ataque) {
        if (new.target === Heroi) {
            throw new Error("Não é possível instanciar a classe abstrata Heroi diretamente.");
        }

        this.nome = nome;
        this.idade = idade;
        this.tipo = tipo;
        this.ataque = ataque;
    }

    atacar() {
        console.log(`O ${this.tipo} atacou usando ${this.ataque}.`);
    }
}

module.exports = Heroi;